In [1]:
import pandas as pd
from sqlalchemy import create_engine

If there will be an error when connecting to server<br>
ModuleNotFoundError: No module named 'psycopg2'<br>
We can try to resolve this by installing psucopg2-binary<br>
Code to run: ''!pip install psycopg2-binary''

# <b><ins>Project description</ins></b> <a class="anchor" id="step0"></a>

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.


# Description of the data

`**books**` - Contains data on books:
- `book_id`
- `author_id`
- `title`
- `num_pages` — number of pages
- `publication_date`
- `publisher_id`
<br>

`**authors**` - Contains data on authors:
- `author_id`
- `author`
<br>

`**publishers**` - Contains data on publishers:
- `publisher_id`
- `publisher`
<br>

`**ratings**` - Contains data on user ratings:
- `rating_id`
- `book_id`
- `username` — the name of the user who rated the book
- `rating`
<br>

`**reviews**` - Contains data on customer reviews:
- `review_id`
- `book_id`
- `username` — the name of the user who reviewed the book
- `text` — the text of the review

# Task

- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
- Find the average number of text reviews among users who rated more than 50 books.

# Instructions for completing the task

- Describe the goals of the study.
- Study the tables (print the first rows).
- Make an SQL query for each of the tasks.
- Output the results of each query in the Notebook.
- Describe your conclusions for each of the tasks.

# Access to the data base

Follow the instructions on connecting to the data base.

- Download the file https://storage.yandexcloud.net/cloud-certs/CA.pem to the folder containing the .ipynb file which you're going to use to work with the data base<br><br>

- Create a data base connection by copying the code below:
```python
    # import libraries
    import pandas as pd
    from sqlalchemy import create_engine
```
```python
    db_config = {'user': 'praktikum_student',         # user name
                 'pwd': 'Sdf4$2;d-d30pp', # password
                 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
                 'port': 6432,              # connection port
                 'db': 'data-analyst-final-project-db'}          # the name of the data base
```
```python
    connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

    engine = create_engine(connection_string, connect_args={'sslmode':'require'})
```
<br>
- The connection is stored in the engine variable. You can run an SQL query using pandas:
```python
    pd.io.sql.read_sql(query, con = engine)
```



# Notes

- Don't forget functions! They can make your life and running queries considerably easier.
- Your results must be obtained with SQL. Use pandas only to print and store query results.

# Data connection

Creating connection to db server

In [2]:
db_config = {'user': 'praktikum_student',                        # user name
             'pwd': 'Sdf4$2;d-d30pp',                            # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,                                       # connection port
             'db': 'data-analyst-final-project-db'}              # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Studying the tables

In next cell I present code that run over all db and print every table that is in "tbl_list" list. <br>
This code is short and not repetitive, but print format is hard readable. So it's better to watch every table separately.

In [3]:
#tbl_list = ['books','authors','publishers','ratings','reviews']

#for tbl in tbl_list:
#    query = ('''SELECT *
#                FROM {}
#                LIMIT 5'''.format(tbl))
#    p = pd.io.sql.read_sql(query, con = engine)
#    print("Table: {}".format(tbl))
#    print(p, '\n')  

In [4]:
books = '''SELECT * FROM books LIMIT 5'''
pd.io.sql.read_sql(books, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
authors = '''SELECT * FROM authors LIMIT 5'''
pd.io.sql.read_sql(authors, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
publishers = '''SELECT * FROM publishers LIMIT 5'''
pd.io.sql.read_sql(publishers, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
ratings = '''SELECT * FROM ratings LIMIT 5'''
pd.io.sql.read_sql(ratings, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
reviews = '''SELECT * FROM reviews LIMIT 5'''
pd.io.sql.read_sql(reviews, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

# Tasks

#### Query 1: Find the number of books released after January 1, 2000.

In [9]:
query_1 = '''
SELECT COUNT (publication_date)
FROM books
WHERE publication_date > '2000-01-01'
'''
books_released = pd.io.sql.read_sql(query_1, con = engine)

p = books_released.values.tolist()
print(p[0])

[819]


As a result, number of books that released after Jan. 1, 2020 equal to 819.

#### Query 2: Find the number of user reviews and the average rating for each book.

In [10]:
query_2 = '''
SELECT books.title as book_title,
    AVG (ratings.rating) as avg_book_rating,
    COUNT (DISTINCT reviews.review_id) AS cnt_of_review
FROM books 
    INNER JOIN ratings ON ratings.book_id = books.book_id 
    INNER JOIN reviews ON reviews.book_id = ratings.book_id
GROUP BY books.book_id
ORDER BY cnt_of_review DESC, avg_book_rating DESC
'''

avg_rating = pd.io.sql.read_sql(query_2, con = engine)
avg_rating.head(10)

book_title  avg_book_rating  \
0                            Twilight (Twilight  #1)         3.662500   
1  Harry Potter and the Prisoner of Azkaban (Harr...         4.414634   
2  Harry Potter and the Chamber of Secrets (Harry...         4.287500   
3                                     The Book Thief         4.264151   
4                                   The Glass Castle         4.206897   
5                The Hobbit  or There and Back Again         4.125000   
6                          Outlander (Outlander  #1)         4.125000   
7  The Curious Incident of the Dog in the Night-Time         4.081081   
8  The Lightning Thief (Percy Jackson and the Oly...         4.080645   
9                                Water for Elephants         3.977273   

   cnt_of_review  
0              7  
1              6  
2              6  
3              6  
4              6  
5              6  
6              6  
7              6  
8              6  
9              6

As a result we can see TOP10 books with the most reviews. <br>
Only one book have maximal reviews number of 7. With average rating 3.66. <br>
Another nine books got 6 reviews each and average rating start from 4.41 until 3.97.

#### Query 3: Identify the publisher that has released the greatest number of books with more than 50 pages.<br> (this will help you exclude brochures and similar publications from your analysis)

In [11]:
query_3 = '''
SELECT publisher AS publisher_name,
    COUNT (books.book_id) AS num_of_books
FROM books 
    INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP  BY publisher
ORDER  BY num_of_books DESC
'''

bst_publisher = pd.io.sql.read_sql(query_3, con = engine)
bst_publisher.head(10)

publisher_name  num_of_books
0             Penguin Books            42
1                   Vintage            31
2  Grand Central Publishing            25
3          Penguin Classics            24
4          Ballantine Books            19
5                    Bantam            19
6                   Berkley            17
7        St. Martin's Press            14
8             Berkley Books            14
9                     Delta            13

The most productive publisher in our data is "Penguin Books" with 42 published books. <br>
All following publishers released less than 31 books.

#### Query 4: Identify the author with the highest average book rating.<br> (look only at books with at least 50 ratings)

In [12]:
query_4 = '''
SELECT joined_table.author, joined_table.cnt_rating, joined_table.avg_book_rating
FROM
(
    SELECT author,
        COUNT (DISTINCT ratings.rating_id) AS cnt_rating,
        AVG (ratings.rating) AS avg_book_rating
    FROM books 
        INNER JOIN authors ON authors.author_id = books.author_id 
        INNER JOIN ratings ON ratings.book_id = books.book_id 
    GROUP BY authors.author_id
    ORDER BY cnt_rating DESC, avg_book_rating DESC
) AS joined_table
WHERE cnt_rating > 50
'''

authors_book_avg_rating = pd.io.sql.read_sql(query_4, con = engine)
authors_book_avg_rating.head(10)

author  cnt_rating  avg_book_rating
0               J.K. Rowling/Mary GrandPré         312         4.288462
1                           J.R.R. Tolkien         166         4.240964
2                          Stephenie Meyer         160         3.662500
3                                Dan Brown         143         3.741259
4                          Nicholas Sparks         111         3.882883
5                             Stephen King         106         4.009434
6                            J.D. Salinger          98         3.846939
7                             Rick Riordan          84         4.130952
8                             Jodi Picoult          76         3.881579
9  George Orwell/Boris Grabnar/Peter Škerl          74         3.729730

The most "rated" author is "J.K. Rowling/Mary GrandPré" with 312 rating marks and average rating of 4.28.

#### Query 5: Find the average number of text reviews among users who rated more than 50 books.

In [13]:
query_5 = '''
SELECT AVG(t2.cnt_text_reviews)

FROM (SELECT username,
      COUNT (DISTINCT ratings.rating_id) AS cnt_num_ratings
      FROM ratings 
      GROUP BY username
      ) AS t1,
      
      (SELECT username,
       COUNT (DISTINCT reviews.review_id) AS cnt_text_reviews
       FROM reviews 
       GROUP BY username
      ) AS t2
      
WHERE t1.username = t2.username 
AND 
cnt_num_ratings > 50

'''

avg_num_reviews = pd.io.sql.read_sql(query_5, con = engine)
avg_num_reviews.head(10)

avg
0  24.333333

The average amount of text reviews between users who rated more than 50 books is 24.33 reviews.
It's almost one review per two rates.